src/amber/controller/helpers/responders.cr

```diff
@@ -61,9 +66,12 @@ module Amber::Controller::Helpers
     end
 
     def set_response(body, status_code = 200, content_type = Content::TYPE[:html])
-      context.response.status_code = status_code
+      if body.is_a?(Proc)
+        body = body.call
+      end
+      context.response.status_code = status_code if context.response.status_code == 200
```

Comment:
`set_response` shouldn't have logic to set the status code or not. IMO any logic on this should be determined on in the `respond_with` method.

`set_response` should just set whatever status_code is passed in or default 200.

```cr
     protected def respond_with(status_code = 200, &block)
       content = with Content.new(requested_responses) yield
       if content.body
         set_response(body: content.body, status_code: status_code, content_type: content.type)
       else
         set_response(body: "Response Not Acceptable.", status_code: 406, content_type: Content::TYPE[:text])
       end
     end
```

Reply:
The reason that check was added to the line that sets the status_code is that a redirect in a Proc may have already set a status_code to 302 - but the respond_with block has a default status_code of 200 which would overwrite the redirect status_code - hence the additional check so that we only use the respond_with status code if the existing status_code is the default 200.  Otherwise the method is as before apart from calling the Proc to get the Proc body content.

Reply:
This could be easily solved by just saying that html views, text, json blocks etc shouldn't be setting response types. We're allowing a block to be passed in so that all response strings don't have to be rendered on load, not so that they can do whatever logic they want.

Reply:
Apart from not rendering something that will be unused my use case is to redirect for one of the formats - e.g. update redirects for html and renders the object for json.  A redirect sets a specific status code.  An alternative would be to set the status as an argument to the format e.g
```
respond_with do
  html 302, ->{redirect("/users"); flash.success("Saved changes to your profile")}
  json user.to_json
end
```
WDYT?